native/explorer/src/series.rs

```diff
@@ -456,6 +456,34 @@ pub fn s_fill_missing_with_bin(data: ExSeries, strategy: &str) -> Result<ExSerie
     Ok(ExSeries::new(s))
 }
 
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn s_fill_missing_with_date(
+    data: ExSeries,
+    strategy: ExDate,
+) -> Result<ExSeries, ExplorerError> {
+    let s = data.clone_inner();
+    let s = s
+        .date()?
+        .fill_null_with_values(strategy.into())?
+        .cast(&DataType::Date)?
+        .into_series();
+    Ok(ExSeries::new(s))
+}
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn s_fill_missing_with_datetime(
+    data: ExSeries,
+    strategy: ExDateTime,
+) -> Result<ExSeries, ExplorerError> {
+    let s = data.clone_inner();
+    let s = s
+        .datetime()?
+        .fill_null_with_values(strategy.into())?
+        .cast(&DataType::Datetime(TimeUnit::Microseconds, None))?
```

Comment:
I don't think we should cast the series here. This will change the whole precision of the series and I am not sure if it is desired. I believe we need to get the precision from the series and convert the datetime to an integer with the same precision. Check encoding.rs, we have some functions that work around this. :)

Reply:
Do all datetime series have the same precision? (As I understand it, these two functions take the precision of the series: [term_from_value](https://github.com/elixir-nx/explorer/blob/main/native/explorer/src/encoding.rs#L389) and [list_from_series](https://github.com/elixir-nx/explorer/blob/main/native/explorer/src/encoding.rs#L403))

If yes, does this code resolve the datetime cast problem?:
```rust
#[rustler::nif(schedule = "DirtyCpu")]
pub fn s_fill_missing_with_datetime(
    data: ExSeries,
    datetime: ExDateTime,
) -> Result<ExSeries, ExplorerError> {
    let s = data.clone_inner();
    let s = s.datetime()?;
    let s = s
        .fill_null_with_values(datetime.into())?
        .cast(&DataType::Datetime(s.time_unit(), None))?
        .into_series();
    Ok(ExSeries::new(s))
}
```

If I understand correctly, if you remove the cast function, the result will already be an integer:
```rust
#[rustler::nif(schedule = "DirtyCpu")]
pub fn s_fill_missing_with_datetime(
    data: ExSeries,
    datetime: ExDateTime,
) -> Result<ExSeries, ExplorerError> {
    let s = data.clone_inner();
    let s = s
        .datetime()?
        .fill_null_with_values(datetime.into())?
        .into_series();
    Ok(ExSeries::new(s))
}
```
```elixir
iex(1)> s1 = Explorer.Series.from_list([~N[2023-01-17 20:00:56.576456], ~N[2023-01-18 20:30:56.576456], nil, ~N[2023-01-09 21:00:56.576456]])
#Explorer.Series<
  Polars[4]
  datetime [2023-01-17 20:00:56.576456, 2023-01-18 20:30:56.576456, nil,
   2023-01-09 21:00:56.576456]
>
iex(2)> Explorer.Series.fill_missing(s1, ~N[2023-01-19 20:00:56.576456])                    
#Explorer.Series<
  Polars[4]
  integer [1673985656576456, 1674073856576456, 1674158456576456,
   1673298056576456]
>
iex(3)> Explorer.Series.fill_missing(s1, ~N[2023-01-19 20:00:56.576456]) |> Explorer.Series.to_list()
[1673985656576456, 1674073856576456, 1674158456576456, 1673298056576456]
```

Reply:
> Do all values ​​in a datetime series have the same precision?

@Jhonatannunessilva I would say yes.

I think if we do as the first example you showed, it's going to be fine.
There is a slightly better improvement: you could take the entire `s.dtype()` to cast back

@josevalim for what I could understand, we need to cast back because `fill_null_with_values` is a function of the `ChunkedArray<Int64>`.

Reply:
Ah, perfect. So casting to s.dtype() sounds the like the right way to go to me! But, if `fill_null_with_values` is expecting an integer as input, then we also need to make sure that the DateTime is cast to the input with the right precision.